Adding datacenter, cluster/ds_cluster labels + some new metrics

dannyk81 opened the pull request:
@pryorda, this fixes https://github.com/pryorda/vmware_exporter/issues/4

This PR implements additional metadata metrics that allows to associate Hosts and VMs to Datacenters and Clusters.

New metrics:
```
vmware_dc_info{dc_name} 1
vmware_dc_cluster_info{dc_name,cluster_name} 1
vmware_dc_cluster_host_info{dc_name,cluster_name,host_name} 1
vmware_dc_cluster_host_vm_info{dc_name,cluster_name,host_name,vm_name} 1
```

Few more minor things:
1) Dockerfile modified to use `ENTRYPOINT` instead of `CMD` so it would be easier to specify a custom config file location.
2) Separated snapshot metrics to its own metrics sub-collection, so these can also be controlled via the `collect_only` section.
3) Few minor log adjustments and re-phrasing

pryorda commented:
Sorry for the delay. I think instead of creating a whole separate metadata metric, we should probably merge them in as labels to vms, snapshots, and hosts?

Let me know your thoughts?

dannyk81 commented:
Thanks for the review @pryorda!

Perhaps we can populate the inventory tree in a dict with `dc_name`, `cluster_name` and `host_name`

We can later use this to lookup the dc and cluster of the VM/Snapshot based on the runtime `host_name` (which we extract from vm summary) and indeed add it as labels.

WDYT?

pryorda commented:
That sounds like a great solution. Let's eating code and then we'll try it out and see if we need to improve on it

dannyk81 commented:
Great 👍 I have it almost ready, just running few tests, will push soon.

dannyk81 commented:
@pryorda first round of changes, last commit removes the metadata metrics and adds `dc_name` and `cluster_name` labels to all metrics except `snapshots` (need some more work there) and datastores.

dannyk81 commented:
some cleanup, and added two host metrics:
```
# HELP vmware_host_maintenance_mode VMWare Host maintenance mode (0 = not in maintenance mode, 1 = in maintenance mode)
# TYPE vmware_host_maintenance_mode gauge
```
```
# HELP vmware_host_connection_state VMWare Host connection state (connected, disconnected or notResponding)
# TYPE vmware_host_connection_state gauge
```

dannyk81 commented:
- improved inventory data collection
- added `dc_name` and `ds_cluster` labels to datastores
- added few more datastore metrics
- switched to `SmartConnect`
- hosts and datastores outside of a cluster will have the `cluster_name` and `ds_cluster` labels empty

pryorda commented:
Looks like the slow down is here, Ignore the ERROR. I need to update that to an info that its likely got a disconnected nic.
```
[2018-10-06 20:19:45.343052+00:00] Total Virtual Machines: 759
[2018-10-06 20:20:02.424743+00:00] Error, cannot get vm metric vmware_vm_net_received_average for old-int-sftp1
[2018-10-06 20:20:02.529178+00:00] Error, cannot get vm metric vmware_vm_net_transmitted_average for old-int-sftp1
[2018-10-06 20:21:46.316558+00:00] Finished VM performance metrics collection
```

pryorda commented:
This diff brought it back down to 59s. Feel free to squash your changes via git rebase then git apply patch.txt and create a final commit. Ill do a one last review once its done. 

[patch.txt](https://github.com/pryorda/vmware_exporter/files/2453492/patch.txt)

```
diff --git a/vmware_exporter/vmware_exporter.py b/vmware_exporter/vmware_exporter.py
index ec1db8d..ed425da 100755
--- a/vmware_exporter/vmware_exporter.py
+++ b/vmware_exporter/vmware_exporter.py
@@ -399,7 +399,7 @@ class VMWareMetricsResource(Resource):
         """
         snapshot_paths = self._vmware_full_snapshots_list(virtual_machine.snapshot.rootSnapshotList)
 
-        host_name, dc_name, cluster_name = self._vmware_vm_metadata(inventory, virtual_machine)
+        _, host_name, dc_name, cluster_name = self._vmware_vm_metadata(inventory, virtual_machine)
 
         for snapshot_path in snapshot_paths:
             snapshot_path['vm_name'] = virtual_machine.name
@@ -513,44 +513,45 @@ class VMWareMetricsResource(Resource):
         vm_power_state = 1 if summary.runtime.powerState == 'poweredOn' else 0
         vm_num_cpu = summary.config.numCpu
 
-        vm_host_name, vm_dc_name, vm_cluster_name = self._vmware_vm_metadata(inventory, virtual_machine)
+        vm_name, vm_host_name, vm_dc_name, vm_cluster_name = self._vmware_vm_metadata(inventory, virtual_machine,
+                                                                                      summary)
+        vm_metadata = [vm_name, vm_host_name, vm_dc_name, vm_cluster_name]
 
         # We gather disk metrics
         if len(virtual_machine.guest.disk) > 0:
             [vm_metrics['vmware_vm_guest_disk_free'].add_metric(
-             [virtual_machine.name, vm_host_name, vm_dc_name, vm_cluster_name, disk.diskPath], disk.freeSpace)
+             [vm_name, vm_host_name, vm_dc_name, vm_cluster_name, disk.diskPath], disk.freeSpace)
              for disk in virtual_machine.guest.disk]
 
             [vm_metrics['vmware_vm_guest_disk_capacity'].add_metric(
-             [virtual_machine.name, vm_host_name, vm_dc_name, vm_cluster_name, disk.diskPath], disk.capacity)
+             [vm_name, vm_host_name, vm_dc_name, vm_cluster_name, disk.diskPath], disk.capacity)
              for disk in virtual_machine.guest.disk]
 
-        vm_metrics['vmware_vm_power_state'].add_metric([virtual_machine.name, vm_host_name,
+        vm_metrics['vmware_vm_power_state'].add_metric([vm_name, vm_host_name,
                                                        vm_dc_name, vm_cluster_name],
                                                        vm_power_state)
-        vm_metrics['vmware_vm_num_cpu'].add_metric([virtual_machine.name, vm_host_name, vm_dc_name, vm_cluster_name],
+        vm_metrics['vmware_vm_num_cpu'].add_metric(vm_metadata,
                                                    vm_num_cpu)
 
         # Get metrics for poweredOn vms only
         if vm_power_state:
             if summary.runtime.bootTime:
                 vm_metrics['vmware_vm_boot_timestamp_seconds'].add_metric(
-                    [virtual_machine.name, vm_host_name, vm_dc_name, vm_cluster_name],
+                    vm_metadata,
                     self._to_epoch(summary.runtime.bootTime))
 
             for p in perf_list:
                 self.threader.thread_it(self._vmware_get_vm_perf_metric,
                                         [content, counter_info, p, virtual_machine,
-                                         vm_metrics, inventory])
+                                         vm_metrics, vm_metadata])
 
         # Debug Me. log("Finished VM: " + vm.name)
 
     def _vmware_get_vm_perf_metric(self, content, counter_info, perf_metric,
-                                   virtual_machine, vm_metrics, inventory):
+                                   virtual_machine, vm_metrics, vm_metadata):
         """
         Get vm perf metric
         """
-        vm_host_name, vm_dc_name, vm_cluster_name = self._vmware_vm_metadata(inventory, virtual_machine)
 
         perf_metric_name = 'vmware_vm_' + perf_metric.replace('.', '_')
         counter_key = counter_info[perf_metric]
@@ -567,11 +568,11 @@ class VMWareMetricsResource(Resource):
         result = content.perfManager.QueryStats(querySpec=[spec])
         # DEBUG ME: log("{0} {1}: {2}".format(vm.name, p, float(sum(result[0].value[0].value))))
         try:
-            vm_metrics[perf_metric_name].add_metric([virtual_machine.name, vm_host_name, vm_dc_name, vm_cluster_name],
+            vm_metrics[perf_metric_name].add_metric(vm_metadata,
                                                     float(sum(result[0].value[0].value)))
         except:  # noqa: E722
             log("Error, cannot get vm metric {0} for {1}".format(perf_metric_name,
-                                                                 virtual_machine.name))
+                                                                 vm_metadata))
 
     def _vmware_get_hosts(self, content, host_metrics, inventory):
         """
@@ -665,18 +666,19 @@ class VMWareMetricsResource(Resource):
 
         return host_inventory, ds_inventory
 
-    def _vmware_vm_metadata(self, inventory, vm):
+    def _vmware_vm_metadata(self, inventory, vm, summary=None):
         """
         Get VM metadata from inventory
         """
-
-        summary = vm.summary
+        if summary is None:
+            summary = vm.summary
+        vm_name = vm.name
         vm_host = summary.runtime.host
         vm_host_name = vm_host.name
         vm_dc_name = inventory[vm_host_name]['dc']
         vm_cluster_name = inventory[vm_host_name]['cluster']
 
-        return vm_host_name, vm_dc_name, vm_cluster_name
+        return vm_name, vm_host_name, vm_dc_name, vm_cluster_name
 
     def _vmware_host_metadata(self, inventory, host):
         """
```

dannyk81 commented:
@pryorda thanks for the final push 😄 

I applied your patch and rebased, PTAL.

pryorda commented:
pryorda/vmware_exporter:v0.2.4 released
